### ISSUE TITLE
[circle-quantizer] Add an option to quantize different ops with different types

### ISSUE THREAD

**Issue by AShedko** [Samsung/ONE#6880]
# What?

Let's allow circle-quantizer to mix quantization types in one model, e. g. the model starts with Q16 quantization and at some point switches to Q8.

# Why?

This can be used for automatically testing downstream tools by generating new models.
Another point is that Q8 is more memory- and compute-efficient than Q16, so it may make sense to only use Q16 when necessary.

# How?

From the user's POV:
A list of pairs `<Op_id, Quantization_type>` is supplied as a command line argument (by filename, for example).
The resulting model's operations have the output types described in the file.

## What's needed from the developer?

Internal changes to circle-quantizer,
parsing the list of pairs,
changes to record-minmax?

**Comment by seanshpark** [comment]
> A list of pairs <Op_id, Quantization_type> is supplied as a command line argument

If there are lots of ops, surely there will be quite a long items.
So I disagree this method.

**Comment by AShedko** [comment]
We can also use `Op_id` ranges, as in `tflitefile_tool/select_operator.py`. Another way is to use some sort of GUI for a user to select subsets of nodes which are then encoded into the list of pairs. We should also allow setting the default quantization type, so
```yaml
default: Q8
0-18: Q16
```
would quantize the first 18 ops as `Q16` and the rest as `Q8`.

**Comment by seanshpark** [comment]
There is a partition tool that has similar issue and this tool uses an external text file, something like a Windows `.ini` file to provide the Ops by type. This still needs revision for better interface.
The GUI you mentioned may produce this file.

Using `0-18: Q16` type is nice but when we think of the worst case (like even node number with Q8 and odd node number with Q16) then there can be lots of arguments.
And getting the node number may give human errors if there are lots of nodes...

**Comment by AShedko** [comment]
Circle nodes also have a name field, so we could use that for selection as well. The selection problem is difficult because we may want all kinds of features, but ultimately the result of parsing is still a list of pairs. IMO, an external tool generating that simple list is the right approach.

**Comment by seanshpark** [comment]
> an external tool generating that simple list is the right approach.

Can you please describe more about this tool? user interface, input of the tool, output of the tool...

**Comment by AShedko** [comment]
The external tool in question, "circle-explorer" should accept as input a circle model and produce as output either the list of pairs for CLI circle-quantizer (useful for automated testing), or directly call circle-quantizer and produce a quantized circle model (may be more ergonomic for end users).

**Comment by seanshpark** [comment]
> The external tool in question, "circle-explorer" should accept as input a circle model and produce as output either the list of pairs for CLI circle-quantizer (useful for automated testing), or directly call circle-quantizer and produce a quantized circle model (may be more ergonomic for end users).

Thanks for the explanation but frankly speaking, I don't understand yet. It's a bit higher level than I expected.

Anyway, it's good to know that you're be working on this topic :)

**Comment by AShedko** [comment]
I have a mock-up UI (in power point -_- ), should I post it here?

**Comment by seanshpark** [comment]
> I have a mock-up UI (in power point -_- ), should I post it here?

Um... I'm not sure as this is open :)
Maybe ask @binarman ?

**Comment by seanshpark** [comment]
I've just reviewed the ppt file :)
GUI tool looks OK. And I think there is some gap between the GUI tool and `circle-quantizer`.
I think the point is that `circle-quantizer` should work alone with selecting Q8-Q16 for the Ops besides the GUI tool.
And it would be better if there are not much burden for the end user working on some huge model...

**Comment by AShedko** [comment]
> I think the point is that circle-quantizer should work alone with selecting Q8-Q16 for the Ops besides the GUI tool.

Yes, that is the idea. GUI is only for convenience. An NN model developer may use their own scripts for generating the inputs to `circle-quantizer`.

**Comment by seanshpark** [comment]
JSON file is used to describe which ops uses what bits and mixed precision is now supported.
